[Windows][dxva] log conversions supported by the processor

CrystalP opened the pull request:
## Description
New function to log the conversion capabilities of a dxva processor, based on the Windows 10 function ID3D11VideoProcessorEnumerator1::CheckVideoProcessorFormatConversion().
On Windows 8 the function exits early and logs nothing.

The input format (NV12, P010, ...) is fixed by the file being played and the function will go over possible combinations of the other 3 parameters of a color conversion:
* input color space (YCbCr only)
* output format (RGB only)
* output color space (RGB only)

In order to make it the logged information more readable the function adds special marks to the values currently chosen by Kodi's algorithms, the values that are a more natural mapping but may be missing processor support,

A few technical points:
* To avoid code duplication, GetDXGIColorSpaceSource() was split in two:
  - color space mapping closest to the description of the media (new function AvToDxgiColorSpace()
  - transform the mappings above to workaround processor limitations and do part of the work in the output stage / shader
* changed the enumerator1 to a class member to avoid creating it in multiple places
* refactored swap chain color spaces enumeration, now also used from outside DeviceResources
* unrelated to PR subject: improve a bit DeviceResources::GetDebugInfo
* also unrelated to PR subject: swapchain3 is enough to call CheckColorSpaceSupport, there was no need for swapchain4,

<!--- Provide a general summary of your change in the Pull Request title above -->
<!--- Describe your change in detail here. -->

## Motivation and context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
Helps with troubleshooting and fixing of dxva processor issues.
A previous PR started testing whether the conversion is supported, this PR lists possible alternatives for a fix.

## How has this been tested?
<!--- Please describe in detail how you tested your change. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
Window 10: Intel 4th and 8th gen, AMD Cezanne igfx
to help with the testing, put a breakpoint in DXVAHD.cpp on the last line of ListSupportedConversions
No Windows 8 needed because the new code exits early under W8.

The processor of Intel 8th gen is strange and CheckVideoProcessorFormatConversion() rejects some conversions even though the processor is able to output a picture (maybe it falls back on a default conversion, with some visual imperfections - unknown at this time). That testing will continue.

## What is the effect on users?
<!--- Summarize the effect of this change on Kodi end-users. -->
<!--- If the PR does not have a noticeable impact (e.g., if it only changes documentation), -->
<!--- just leave it empty. Put in more detail the bigger the impact is. -->
<!--- This section may be used for automatic creation of release notes. -->
no direct effect and the code activates only when the log level is debug.

## Screenshots (if appropriate):
Example of a couple lines - not a complete log

![image](https://user-images.githubusercontent.com/489377/231375775-6e9d3e25-3608-4b60-a772-fa80ab974128.png)

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [X] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

thexai commented:
The work and effort in making this is appreciated but there are some aspects of the design that I don't like:

Too many changes are made to the current code (which works fine) and therefore may introduce bugs. The changes do not seem clear to me and the code is difficult to review. There are also changes that do not seem necessary or are not directly related to the PR.

First I would ask to separate the deviceResources changes in a separate commit. Perhaps the changes in debug info are unnecessary and penalize performance (it is not necessary to evaluate 100 pixel formats if swap chain can only have 2). Note that this code is executed 60 times per second while the debug OSD is visible.

I don't think it's necessary to make changes to `GetDXGIColorSpaceSource` since according to the PR it only makes a detailed log of the possible combinations supported but should not change (either intentionally or by mistake) the current operation.

That is, the new log function should call `GetDXGIColorSpaceSource` as it currently is and implement additional code if necessary without changing the existing one.

I also don't think there is much value in listing combinations that are by nature impossible i.e. all X601, P601, P709 color space for videos that are BT.2020:

```
     (*: values picked by heuristics, N native input color space, bb supported as swap chain backbuffer)
     * P010 /    YCBCR_FULL_G22_NONE_P709_X601    to   R16G16B16A16_FLOAT         / *bb RGB_FULL_G22_NONE_P709          
     * P010 /    YCBCR_FULL_G22_NONE_P709_X601    to   R16G16B16A16_FLOAT         /     RGB_FULL_G10_NONE_P709          
     * P010 /    YCBCR_FULL_G22_NONE_P709_X601    to   R16G16B16A16_FLOAT         /  bb RGB_STUDIO_G22_NONE_P709        
     * P010 /    YCBCR_FULL_G22_NONE_P709_X601    to * R10G10B10A2_UNORM          / *bb RGB_FULL_G22_NONE_P709          
     * P010 /    YCBCR_FULL_G22_NONE_P709_X601    to * R10G10B10A2_UNORM          /     RGB_FULL_G10_NONE_P709          
     * P010 /    YCBCR_FULL_G22_NONE_P709_X601    to * R10G10B10A2_UNORM          /  bb RGB_STUDIO_G22_NONE_P709        
     * P010 /    YCBCR_FULL_G22_NONE_P709_X601    to   R8G8B8A8_UNORM             / *bb RGB_FULL_G22_NONE_P709          
     * P010 /    YCBCR_FULL_G22_NONE_P709_X601    to   R8G8B8A8_UNORM             /     RGB_FULL_G10_NONE_P709          
     * P010 /    YCBCR_FULL_G22_NONE_P709_X601    to   R8G8B8A8_UNORM             /  bb RGB_STUDIO_G22_NONE_P709        
     * P010 /    YCBCR_FULL_G22_NONE_P709_X601    to   B8G8R8A8_UNORM             / *bb RGB_FULL_G22_NONE_P709          
     * P010 /    YCBCR_FULL_G22_NONE_P709_X601    to   B8G8R8A8_UNORM             /     RGB_FULL_G10_NONE_P709          
     * P010 /    YCBCR_FULL_G22_NONE_P709_X601    to   B8G8R8A8_UNORM             /  bb RGB_STUDIO_G22_NONE_P709        
     * P010 /    YCBCR_FULL_G22_NONE_P709_X601    to   B8G8R8X8_UNORM             / *bb RGB_FULL_G22_NONE_P709          
     * P010 /    YCBCR_FULL_G22_NONE_P709_X601    to   B8G8R8X8_UNORM             /     RGB_FULL_G10_NONE_P709          
     * P010 /    YCBCR_FULL_G22_NONE_P709_X601    to   B8G8R8X8_UNORM             /  bb RGB_STUDIO_G22_NONE_P709        
     * P010 /    YCBCR_STUDIO_G22_LEFT_P601       to   R16G16B16A16_FLOAT         / *bb RGB_FULL_G22_NONE_P709          
     * P010 /    YCBCR_STUDIO_G22_LEFT_P601       to   R16G16B16A16_FLOAT         /     RGB_FULL_G10_NONE_P709          
     * P010 /    YCBCR_STUDIO_G22_LEFT_P601       to   R16G16B16A16_FLOAT         /  bb RGB_STUDIO_G22_NONE_P709        
     * P010 /    YCBCR_STUDIO_G22_LEFT_P601       to * R10G10B10A2_UNORM          / *bb RGB_FULL_G22_NONE_P709          
     * P010 /    YCBCR_STUDIO_G22_LEFT_P601       to * R10G10B10A2_UNORM          /     RGB_FULL_G10_NONE_P709          
     * P010 /    YCBCR_STUDIO_G22_LEFT_P601       to * R10G10B10A2_UNORM          /  bb RGB_STUDIO_G22_NONE_P709        
     * P010 /    YCBCR_STUDIO_G22_LEFT_P601       to   R8G8B8A8_UNORM             / *bb RGB_FULL_G22_NONE_P709          
     * P010 /    YCBCR_STUDIO_G22_LEFT_P601       to   R8G8B8A8_UNORM             /     RGB_FULL_G10_NONE_P709          
     * P010 /    YCBCR_STUDIO_G22_LEFT_P601       to   R8G8B8A8_UNORM             /  bb RGB_STUDIO_G22_NONE_P709        
     * P010 /    YCBCR_STUDIO_G22_LEFT_P601       to   B8G8R8A8_UNORM             / *bb RGB_FULL_G22_NONE_P709          
     * P010 /    YCBCR_STUDIO_G22_LEFT_P601       to   B8G8R8A8_UNORM             /     RGB_FULL_G10_NONE_P709          
     * P010 /    YCBCR_STUDIO_G22_LEFT_P601       to   B8G8R8A8_UNORM             /  bb RGB_STUDIO_G22_NONE_P709        
     * P010 /    YCBCR_STUDIO_G22_LEFT_P601       to   B8G8R8X8_UNORM             / *bb RGB_FULL_G22_NONE_P709          
     * P010 /    YCBCR_STUDIO_G22_LEFT_P601       to   B8G8R8X8_UNORM             /     RGB_FULL_G10_NONE_P709          
     * P010 /    YCBCR_STUDIO_G22_LEFT_P601       to   B8G8R8X8_UNORM             /  bb RGB_STUDIO_G22_NONE_P709        
     * P010 /    YCBCR_FULL_G22_LEFT_P601         to   R16G16B16A16_FLOAT         / *bb RGB_FULL_G22_NONE_P709          
     * P010 /    YCBCR_FULL_G22_LEFT_P601         to   R16G16B16A16_FLOAT         /     RGB_FULL_G10_NONE_P709          
     * P010 /    YCBCR_FULL_G22_LEFT_P601         to   R16G16B16A16_FLOAT         /  bb RGB_STUDIO_G22_NONE_P709        
     * P010 /    YCBCR_FULL_G22_LEFT_P601         to * R10G10B10A2_UNORM          / *bb RGB_FULL_G22_NONE_P709          
     * P010 /    YCBCR_FULL_G22_LEFT_P601         to * R10G10B10A2_UNORM          /     RGB_FULL_G10_NONE_P709          
     * P010 /    YCBCR_FULL_G22_LEFT_P601         to * R10G10B10A2_UNORM          /  bb RGB_STUDIO_G22_NONE_P709        
     * P010 /    YCBCR_FULL_G22_LEFT_P601         to   R8G8B8A8_UNORM             / *bb RGB_FULL_G22_NONE_P709          
     * P010 /    YCBCR_FULL_G22_LEFT_P601         to   R8G8B8A8_UNORM             /     RGB_FULL_G10_NONE_P709          
     * P010 /    YCBCR_FULL_G22_LEFT_P601         to   R8G8B8A8_UNORM             /  bb RGB_STUDIO_G22_NONE_P709        
     * P010 /    YCBCR_FULL_G22_LEFT_P601         to   B8G8R8A8_UNORM             / *bb RGB_FULL_G22_NONE_P709          
     * P010 /    YCBCR_FULL_G22_LEFT_P601         to   B8G8R8A8_UNORM             /     RGB_FULL_G10_NONE_P709          
     * P010 /    YCBCR_FULL_G22_LEFT_P601         to   B8G8R8A8_UNORM             /  bb RGB_STUDIO_G22_NONE_P709        
     * P010 /    YCBCR_FULL_G22_LEFT_P601         to   B8G8R8X8_UNORM             / *bb RGB_FULL_G22_NONE_P709          
     * P010 /    YCBCR_FULL_G22_LEFT_P601         to   B8G8R8X8_UNORM             /     RGB_FULL_G10_NONE_P709          
     * P010 /    YCBCR_FULL_G22_LEFT_P601         to   B8G8R8X8_UNORM             /  bb RGB_STUDIO_G22_NONE_P709        
     * P010 /    YCBCR_STUDIO_G22_LEFT_P709       to   R16G16B16A16_FLOAT         / *bb RGB_FULL_G22_NONE_P709          
     * P010 /    YCBCR_STUDIO_G22_LEFT_P709       to   R16G16B16A16_FLOAT         /     RGB_FULL_G10_NONE_P709          
     * P010 /    YCBCR_STUDIO_G22_LEFT_P709       to   R16G16B16A16_FLOAT         /  bb RGB_STUDIO_G22_NONE_P709        
     * P010 /    YCBCR_STUDIO_G22_LEFT_P709       to * R10G10B10A2_UNORM          / *bb RGB_FULL_G22_NONE_P709          
     * P010 /    YCBCR_STUDIO_G22_LEFT_P709       to * R10G10B10A2_UNORM          /     RGB_FULL_G10_NONE_P709          
     * P010 /    YCBCR_STUDIO_G22_LEFT_P709       to * R10G10B10A2_UNORM          /  bb RGB_STUDIO_G22_NONE_P709        
     * P010 /    YCBCR_STUDIO_G22_LEFT_P709       to   R8G8B8A8_UNORM             / *bb RGB_FULL_G22_NONE_P709          
     * P010 /    YCBCR_STUDIO_G22_LEFT_P709       to   R8G8B8A8_UNORM             /     RGB_FULL_G10_NONE_P709          
     * P010 /    YCBCR_STUDIO_G22_LEFT_P709       to   R8G8B8A8_UNORM             /  bb RGB_STUDIO_G22_NONE_P709        
     * P010 /    YCBCR_STUDIO_G22_LEFT_P709       to   B8G8R8A8_UNORM             / *bb RGB_FULL_G22_NONE_P709          
     * P010 /    YCBCR_STUDIO_G22_LEFT_P709       to   B8G8R8A8_UNORM             /     RGB_FULL_G10_NONE_P709          
     * P010 /    YCBCR_STUDIO_G22_LEFT_P709       to   B8G8R8A8_UNORM             /  bb RGB_STUDIO_G22_NONE_P709        
     * P010 /    YCBCR_STUDIO_G22_LEFT_P709       to   B8G8R8X8_UNORM             / *bb RGB_FULL_G22_NONE_P709          
     * P010 /    YCBCR_STUDIO_G22_LEFT_P709       to   B8G8R8X8_UNORM             /     RGB_FULL_G10_NONE_P709          
     * P010 /    YCBCR_STUDIO_G22_LEFT_P709       to   B8G8R8X8_UNORM             /  bb RGB_STUDIO_G22_NONE_P709        
     * P010 /    YCBCR_FULL_G22_LEFT_P709         to   R16G16B16A16_FLOAT         / *bb RGB_FULL_G22_NONE_P709          
     * P010 /    YCBCR_FULL_G22_LEFT_P709         to   R16G16B16A16_FLOAT         /     RGB_FULL_G10_NONE_P709          
     * P010 /    YCBCR_FULL_G22_LEFT_P709         to   R16G16B16A16_FLOAT         /  bb RGB_STUDIO_G22_NONE_P709        
     * P010 /    YCBCR_FULL_G22_LEFT_P709         to * R10G10B10A2_UNORM          / *bb RGB_FULL_G22_NONE_P709          
     * P010 /    YCBCR_FULL_G22_LEFT_P709         to * R10G10B10A2_UNORM          /     RGB_FULL_G10_NONE_P709          
     * P010 /    YCBCR_FULL_G22_LEFT_P709         to * R10G10B10A2_UNORM          /  bb RGB_STUDIO_G22_NONE_P709        
     * P010 /    YCBCR_FULL_G22_LEFT_P709         to   R8G8B8A8_UNORM             / *bb RGB_FULL_G22_NONE_P709          
     * P010 /    YCBCR_FULL_G22_LEFT_P709         to   R8G8B8A8_UNORM             /     RGB_FULL_G10_NONE_P709          
     * P010 /    YCBCR_FULL_G22_LEFT_P709         to   R8G8B8A8_UNORM             /  bb RGB_STUDIO_G22_NONE_P709        
     * P010 /    YCBCR_FULL_G22_LEFT_P709         to   B8G8R8A8_UNORM             / *bb RGB_FULL_G22_NONE_P709          
     * P010 /    YCBCR_FULL_G22_LEFT_P709         to   B8G8R8A8_UNORM             /     RGB_FULL_G10_NONE_P709          
     * P010 /    YCBCR_FULL_G22_LEFT_P709         to   B8G8R8A8_UNORM             /  bb RGB_STUDIO_G22_NONE_P709        
     * P010 /    YCBCR_FULL_G22_LEFT_P709         to   B8G8R8X8_UNORM             / *bb RGB_FULL_G22_NONE_P709          
     * P010 /    YCBCR_FULL_G22_LEFT_P709         to   B8G8R8X8_UNORM             /     RGB_FULL_G10_NONE_P709          
     * P010 /    YCBCR_FULL_G22_LEFT_P709         to   B8G8R8X8_UNORM             /  bb RGB_STUDIO_G22_NONE_P709        
     * P010 / *  YCBCR_STUDIO_G22_LEFT_P2020      to   R16G16B16A16_FLOAT         / *bb RGB_FULL_G22_NONE_P709          
     * P010 / *  YCBCR_STUDIO_G22_LEFT_P2020      to   R16G16B16A16_FLOAT         /  bb RGB_STUDIO_G22_NONE_P709        
     * P010 / *  YCBCR_STUDIO_G22_LEFT_P2020      to * R10G10B10A2_UNORM          / *bb RGB_FULL_G22_NONE_P709          
     * P010 / *  YCBCR_STUDIO_G22_LEFT_P2020      to * R10G10B10A2_UNORM          /  bb RGB_STUDIO_G22_NONE_P709        
     * P010 / *  YCBCR_STUDIO_G22_LEFT_P2020      to   R8G8B8A8_UNORM             / *bb RGB_FULL_G22_NONE_P709          
     * P010 / *  YCBCR_STUDIO_G22_LEFT_P2020      to   R8G8B8A8_UNORM             /  bb RGB_STUDIO_G22_NONE_P709        
     * P010 / *  YCBCR_STUDIO_G22_LEFT_P2020      to   B8G8R8A8_UNORM             / *bb RGB_FULL_G22_NONE_P709          
     * P010 / *  YCBCR_STUDIO_G22_LEFT_P2020      to   B8G8R8A8_UNORM             /  bb RGB_STUDIO_G22_NONE_P709        
     * P010 / *  YCBCR_STUDIO_G22_LEFT_P2020      to   B8G8R8X8_UNORM             / *bb RGB_FULL_G22_NONE_P709          
     * P010 / *  YCBCR_STUDIO_G22_LEFT_P2020      to   B8G8R8X8_UNORM             /  bb RGB_STUDIO_G22_NONE_P709        
     * P010 /  N YCBCR_STUDIO_G2084_LEFT_P2020    to   R16G16B16A16_FLOAT         /     RGB_FULL_G10_NONE_P709          
     * P010 /  N YCBCR_STUDIO_G2084_LEFT_P2020    to   R16G16B16A16_FLOAT         /  bb RGB_FULL_G2084_NONE_P2020       
     * P010 /  N YCBCR_STUDIO_G2084_LEFT_P2020    to * R10G10B10A2_UNORM          /     RGB_FULL_G10_NONE_P709          
     * P010 /  N YCBCR_STUDIO_G2084_LEFT_P2020    to * R10G10B10A2_UNORM          /  bb RGB_FULL_G2084_NONE_P2020       
     * P010 /  N YCBCR_STUDIO_G2084_LEFT_P2020    to   R8G8B8A8_UNORM             /     RGB_FULL_G10_NONE_P709          
     * P010 /  N YCBCR_STUDIO_G2084_LEFT_P2020    to   R8G8B8A8_UNORM             /  bb RGB_FULL_G2084_NONE_P2020       
     * P010 /  N YCBCR_STUDIO_G2084_LEFT_P2020    to   B8G8R8A8_UNORM             /     RGB_FULL_G10_NONE_P709          
     * P010 /  N YCBCR_STUDIO_G2084_LEFT_P2020    to   B8G8R8A8_UNORM             /  bb RGB_FULL_G2084_NONE_P2020       
     * P010 /  N YCBCR_STUDIO_G2084_LEFT_P2020    to   B8G8R8X8_UNORM             /     RGB_FULL_G10_NONE_P709          
     * P010 /  N YCBCR_STUDIO_G2084_LEFT_P2020    to   B8G8R8X8_UNORM             /  bb RGB_FULL_G2084_NONE_P2020       
     * P010 /    YCBCR_STUDIO_G2084_TOPLEFT_P2020 to   R16G16B16A16_FLOAT         /     RGB_FULL_G10_NONE_P709          
     * P010 /    YCBCR_STUDIO_G2084_TOPLEFT_P2020 to   R16G16B16A16_FLOAT         /  bb RGB_FULL_G2084_NONE_P2020       
     * P010 /    YCBCR_STUDIO_G2084_TOPLEFT_P2020 to * R10G10B10A2_UNORM          /     RGB_FULL_G10_NONE_P709          
     * P010 /    YCBCR_STUDIO_G2084_TOPLEFT_P2020 to * R10G10B10A2_UNORM          /  bb RGB_FULL_G2084_NONE_P2020       
     * P010 /    YCBCR_STUDIO_G2084_TOPLEFT_P2020 to   R8G8B8A8_UNORM             /     RGB_FULL_G10_NONE_P709          
     * P010 /    YCBCR_STUDIO_G2084_TOPLEFT_P2020 to   R8G8B8A8_UNORM             /  bb RGB_FULL_G2084_NONE_P2020       
     * P010 /    YCBCR_STUDIO_G2084_TOPLEFT_P2020 to   B8G8R8A8_UNORM             /     RGB_FULL_G10_NONE_P709          
     * P010 /    YCBCR_STUDIO_G2084_TOPLEFT_P2020 to   B8G8R8A8_UNORM             /  bb RGB_FULL_G2084_NONE_P2020       
     * P010 /    YCBCR_STUDIO_G2084_TOPLEFT_P2020 to   B8G8R8X8_UNORM             /     RGB_FULL_G10_NONE_P709          
     * P010 /    YCBCR_STUDIO_G2084_TOPLEFT_P2020 to   B8G8R8X8_UNORM             /  bb RGB_FULL_G2084_NONE_P2020   
```

Perhaps it would be better to limit the listing to color spaces that match the source only.

CrystalP commented:
No worries, criticize away, there are many ways to do this and to present the information, it took me a while to settle on something but still am not very happy. I wish that information was dumped by dxdiag and this whole thing could be avoided, but it's not.

I'm going to split into multiple commits.

- Useless input color spaces
It's not that easy to define and code "common sense", filter out some lines but keep others that may or may not be useful in the future.  The function exists to help with future unknown situations (missing conversion) or optimize some choices.
With those parameters, might as well use the dumb solution of listing all and avoid complexity, extra code, enums, mappings...
For example the HDR>SDR tonemapping path overrides the "common sense" choice of input color space (ie YCBCR_STUDIO_G2084_LEFT_P2020  or the TOPLEFT). That was not obvious. Same for HLG in HDR.

The input format is not very helpful to cut the list. P010 does not automatically mean BT2020. Encoding SDR with 10 bit H265 is common, same for anime in 10 bit H264.
I have one example of the opposite, G2084/BT2020 in NV12 though it's probably an exception...

- GetDXGIColorSpaceSource
I needed a function that provides the "naive" mapping from the Av* fields to a DXGI color space > a generic AvToDxgiColorSpace was created.
It is quite similar to GetDXGIColorSpaceSource, except for the cases that apply workarounds for various limitations of processors. The idea was to keep in GetDXGIColorSpaceSource only the code for the differences from the ideal situation of a perfect processor, but the more workarounds there are, the more complicated the logic gets (not a good thing, but hidden with the way GetDXGIColorSpaceSource is today).
GetDXGIColorSpaceSource could remain unchanged, not a problem.

- no problem reverting the lookup in the video debug OSD (DeviceResources), it made sense when I had std::map containing the texts at one point but not anymore. I'll put the flag testing in a separate commit as suggested, it's still valid I think.

CrystalP committed:
[Windows] refactor listing of swap chain formats

CrystalP committed:
[Windows] fix test of DXGI_SWAP_CHAIN_DESC1.Flags for fullscreen exclusive determination

CrystalP commented:
Here it is, broken up in discrete steps that should be easier to follow.

Tested with SDR/HDR/HLG clips played as HDR and SDR, with HDR capable and not capable hardware.
AMD Cezanne iGPU, Intel 4th and 8th gen.

Functionality is the same as before PR in GetDXGIColorSpaceSource, with the small exception of doing for HLG what you added in 5b528b9665e4e1f2d89d10756b38f038a63041f4 / PR#23109 (processor supporting TOPLEFT not LEFT)

To me the benefit of separating standard mapping from the various processor limitation workarounds is that it's much easier to keep track of them that way and it should be easier when more have to be added.
If the list grows it will be possible to refactor the "override" functions to have an identical signature, define the list of override functions in an array and simply loop over the array, for any number of them.

CrystalP commented:
I had already corrected and pushed for Jenkins last diff suggestion - not applicable.
The diffs and github presentation are making a mess of the GetDXGIColorSpaceSource work because it finds some similar lines, but it would be better to compare full function before and after, not bits and pieces of it.

thexai commented on xbmc/rendering/dx/DeviceResources.cpp:
```diff
@@ -716,26 +716,14 @@ void DX::DeviceResources::ResizeBuffers()
     {
-      ComPtr<IDXGISwapChain4> swapChain4;
-      if (SUCCEEDED(m_swapChain.As(&swapChain4)))
+      std::vector<DXGI_COLOR_SPACE_TYPE> cs = SwapChainColorSpaces();
```
This variable `cs` is not really necessary. You can can directly `SwapChainColorSpaces` inside for loop:

`for (const auto& colorSpace : SwapChainColorSpaces())`

CrystalP replied:
yes

CrystalP commented on xbmc/cores/VideoPlayer/VideoRenderers/HwDecRender/DXVAHD.cpp:
```diff
@@ -811,3 +812,209 @@ bool CProcessorHD::IsBT2020Supported()
+          return DXGI_COLOR_SPACE_RGB_STUDIO_G2084_NONE_P2020;
+
+        return DXGI_COLOR_SPACE_RGB_STUDIO_G22_NONE_P2020;
+      }
```
@thexai in the same place, GetDXGIColorSpaceSource has an override to G22 when !supportHDR:

```
        if (csArgs.color_transfer == AVCOL_TRC_SMPTEST2084 && supportHDR)
          return DXGI_COLOR_SPACE_RGB_STUDIO_G2084_NONE_P2020;

        return DXGI_COLOR_SPACE_RGB_STUDIO_G22_NONE_P2020;

```
That was added with HDR support. But is there any way that the processor will ever see RGB input? Is the rest of the code actually ready for this?

thexai replied:
As commented the other day changes to `GetDXGIColorSpaceSource` seems not relevant for this PR aka "log conversions supported by the processor"

I would like to be able to separate these changes into another PR. I'm concerned about performance issues as these functions are called for each rendered frame 24 or 60 times / second and really have nothing to do with this PR (log conversions supported).

CrystalP replied:
yes, I'll force push. if you imported the branch you only have to pop the last two commits.

perf should be pretty close but I have other ideas to side step the issue completely.

thexai commented on xbmc/rendering/dx/DeviceResources.cpp:
```diff
@@ -716,26 +716,14 @@ void DX::DeviceResources::ResizeBuffers()
-      ComPtr<IDXGISwapChain4> swapChain4;
-      if (SUCCEEDED(m_swapChain.As(&swapChain4)))
+      std::vector<DXGI_COLOR_SPACE_TYPE> cs = SwapChainColorSpaces();
+      std::string colorSpaces{};
```
Not need initialize `std::string` with {}

CrystalP replied:
yes

thexai commented on xbmc/rendering/dx/DeviceResources.cpp:
```diff
@@ -716,26 +716,14 @@ void DX::DeviceResources::ResizeBuffers()
-      if (SUCCEEDED(m_swapChain.As(&swapChain4)))
+      std::vector<DXGI_COLOR_SPACE_TYPE> cs = SwapChainColorSpaces();
+      std::string colorSpaces{};
+      for (auto colorSpace : cs)
```
`const auto&`

CrystalP replied:
it's a value type, seems overkill?

thexai commented on xbmc/rendering/dx/DeviceResources.cpp:
```diff
@@ -716,26 +716,14 @@ void DX::DeviceResources::ResizeBuffers()
-        }
-        CLog::LogF(LOGDEBUG, "Color spaces supported by the swap chain:{}", colorSpaces);
+        colorSpaces.append("\n");
+        colorSpaces.append(DX::DXGIColorSpaceTypeToString(colorSpace));
```
Is possible ovoid two .append() calls with:

colorSpaces.append('\n' + DX::DXGIColorSpaceTypeToString(colorSpace));

CrystalP replied:
yes but you end up building more strings that way?

thexai commented on xbmc/rendering/dx/DeviceResources.cpp:
```diff
@@ -1399,3 +1387,29 @@ DEBUG_INFO_RENDER DX::DeviceResources::GetDebugInfo() const
+  if (SUCCEEDED(hr = m_swapChain.As(&swapChain3)))
+  {
+    UINT colorSpaceSupport = 0;
+    std::string colorSpaces{};
```
Unused variable `colorSpaces` ?

CrystalP replied:
yes, leftover from copy/paste

thexai commented on xbmc/rendering/dx/DeviceResources.cpp:
```diff
@@ -1399,3 +1386,28 @@ DEBUG_INFO_RENDER DX::DeviceResources::GetDebugInfo() const
   return info;
 }
+
+std::vector<DXGI_COLOR_SPACE_TYPE> DX::DeviceResources::SwapChainColorSpaces()
```
```suggestion
std::vector<DXGI_COLOR_SPACE_TYPE> DX::DeviceResources::SwapChainColorSpaces() const
```

CrystalP replied:
ok

thexai commented on xbmc/rendering/dx/DeviceResources.cpp:
```diff
@@ -716,26 +716,13 @@ void DX::DeviceResources::ResizeBuffers()
-      ComPtr<IDXGISwapChain4> swapChain4;
-      if (SUCCEEDED(m_swapChain.As(&swapChain4)))
+      std::string colorSpaces;
+      for (DXGI_COLOR_SPACE_TYPE colorSpace : SwapChainColorSpaces())
```
```suggestion
      for (const DXGI_COLOR_SPACE_TYPE& colorSpace : SwapChainColorSpaces())
```

or `const auto&`

CrystalP replied:
ok

thexai commented on xbmc/cores/VideoPlayer/VideoRenderers/HwDecRender/DXVAHD.cpp:
```diff
@@ -259,33 +262,31 @@ bool CProcessorHD::InitProcessor()
-  ComPtr<ID3D11VideoProcessorEnumerator1> pEnumerator1;
-
-  if (SUCCEEDED(m_pEnumerator.As(&pEnumerator1)))
+  if (SUCCEEDED(m_pEnumerator.As(&m_pEnumerator1)))
```
I'm not sure about this change because processor is Re-Init when is lost and is used `pEnumerator` to track if is valid: 

https://github.com/xbmc/xbmc/blob/a3e7d1a9b0f3da68f1571eb3981e74920f536a2f/xbmc/cores/VideoPlayer/VideoRenderers/HwDecRender/DXVAHD.cpp#L383-L403

If we use an old instance of  `m_pEnumerator1` maybe is invalid reference or cause deadlock, etc. Think is more safe at it is now.

CrystalP replied:
I made sure it's constructed / destructed in sync with the existing m_pEnumerator that it is based on. It's protected by the same locks, so seems to me we won't have situation where one is valid and not the other one (under Windows 10 with API is available).
Still, it's possible to move the creation a bit earlier so there are no return statements between m_pEnumerator  creation and m_pEnumerator1 creation.
Also possible to take the lock when entering IsFormatConversionSupported and ListSupportedConversions, called from out of class.

thexai commented on xbmc/cores/VideoPlayer/VideoRenderers/HwDecRender/DXVAHD.cpp:
```diff
@@ -814,3 +812,209 @@ bool CProcessorHD::IsBT2020Supported()
+  if (!m_pEnumerator1)
+    return;
+
+  DXGIColorSpaceArgs csArgs = DXGIColorSpaceArgs(picture);
```
```suggestion
  const DXGIColorSpaceArgs csArgs = DXGIColorSpaceArgs(picture);
```

CrystalP replied:
yes to all const and const &  comments

thexai commented on xbmc/cores/VideoPlayer/VideoRenderers/HwDecRender/DXVAHD.cpp:
```diff
@@ -814,3 +812,209 @@ bool CProcessorHD::IsBT2020Supported()
+  DXGIColorSpaceArgs csArgs = DXGIColorSpaceArgs(picture);
+
+  // Defaults used by Kodi
+  ProcColorSpaces heuristicsCS = CalculateDXGIColorSpaces(csArgs);
```
```suggestion
  const ProcColorSpaces heuristicsCS = CalculateDXGIColorSpaces(csArgs);
```

thexai commented on xbmc/cores/VideoPlayer/VideoRenderers/HwDecRender/DXVAHD.cpp:
```diff
@@ -814,3 +812,209 @@ bool CProcessorHD::IsBT2020Supported()
+    return;
+  }
+
+  DXGI_COLOR_SPACE_TYPE inputNativeCS = AvToDxgiColorSpace(csArgs);
```
```suggestion
  const DXGI_COLOR_SPACE_TYPE inputNativeCS = AvToDxgiColorSpace(csArgs);
```

thexai commented on xbmc/cores/VideoPlayer/VideoRenderers/HwDecRender/DXVAHD.cpp:
```diff
@@ -814,3 +812,209 @@ bool CProcessorHD::IsBT2020Supported()
+  }
+
+  // Only probe the output formats of RGB/BGR type supported by the processor.
+  std::vector<DXGI_FORMAT> allOutputFormats = ProcessorOutputFormats();
```
Maybe use directly in for loop:
```
for (const auto& format : ProcessorOutputFormats())
{
```

CrystalP replied:
definitely

thexai commented on xbmc/cores/VideoPlayer/VideoRenderers/HwDecRender/DXVAHD.cpp:
```diff
@@ -814,3 +812,209 @@ bool CProcessorHD::IsBT2020Supported()
+  // Only probe the output formats of RGB/BGR type supported by the processor.
+  std::vector<DXGI_FORMAT> allOutputFormats = ProcessorOutputFormats();
+  std::vector<DXGI_FORMAT> outputFormats;
+  for (auto format : allOutputFormats)
```
const auto&

thexai commented on xbmc/cores/VideoPlayer/VideoRenderers/HwDecRender/DXVAHD.cpp:
```diff
@@ -814,3 +812,209 @@ bool CProcessorHD::IsBT2020Supported()
+  // - input color spaces
+  // - output formats
+  // - output color spaces
+  for (auto inputCS : ycbcrColorSpaces)
```
```suggestion
  for (const auto& inputCS : ycbcrColorSpaces)
```

thexai commented on xbmc/cores/VideoPlayer/VideoRenderers/HwDecRender/DXVAHD.cpp:
```diff
@@ -814,3 +812,209 @@ bool CProcessorHD::IsBT2020Supported()
+  // - output color spaces
+  for (auto inputCS : ycbcrColorSpaces)
+  {
+    for (DXGI_FORMAT outputFormat : outputFormats)
```
```suggestion
    for (const DXGI_FORMAT& outputFormat : outputFormats)
```

thexai commented on xbmc/cores/VideoPlayer/VideoRenderers/HwDecRender/DXVAHD.cpp:
```diff
@@ -814,3 +812,209 @@ bool CProcessorHD::IsBT2020Supported()
+  {
+    for (DXGI_FORMAT outputFormat : outputFormats)
+    {
+      for (DXGI_COLOR_SPACE_TYPE outputCS : rgbColorSpaces)
```
```suggestion
      for (const DXGI_COLOR_SPACE_TYPE& outputCS : rgbColorSpaces)
```

thexai commented on xbmc/cores/VideoPlayer/VideoRenderers/HwDecRender/DXVAHD.h:
```diff
@@ -132,6 +141,8 @@ class CProcessorHD : public ID3DResource
                                                 bool supportHDR,
                                                 bool limitedRange) const;
+  static DXGI_COLOR_SPACE_TYPE AvToDxgiColorSpace(const DXGIColorSpaceArgs& csArgs);
+  std::vector<DXGI_FORMAT> ProcessorOutputFormats() const;
```
Maybe rename to `GetProcessorOutputFormats`

CrystalP replied:
good idea

thexai commented on xbmc/cores/VideoPlayer/VideoRenderers/HwDecRender/DXVAHD.cpp:
```diff
@@ -814,3 +813,208 @@ bool CProcessorHD::IsBT2020Supported()
+  UINT uiFlags;
+  for (int fmt = DXGI_FORMAT_UNKNOWN; fmt <= DXGI_FORMAT_V408; fmt++)
+  {
+    if (S_OK == m_pEnumerator->CheckVideoProcessorFormat((DXGI_FORMAT)fmt, &uiFlags) &&
```
Avoid C style cast, maybe create aux variable as is used later in `push_back()`

`DXGI_FORMAT format = static_cast<DXGI_FORMAT>(fmt);`

CrystalP commented:
Feedback merged
- const / const&
- renamed ProcessorOutputFormats > GetProcessorOutputFormats
- did the same for SwapChainColorSpaces > GetSwapChainColorSpaces of commit 1
- moved up creation of enumerator1 right after creation of enumerator
- take the lock when entering  ListSupportedConversions() or IsFormatConversionSupported() (had to remove const)

- protected IsFormatConversionSupported() against non-existing enumerator1
- also moved up a couple lines in ListSupportedConversion that can be executed in Windows 8 with basic enumerator

I can't say I understand the threading model of Kodi and whether it's a possible scenario, but now the enumerator/enumerator1 shouldn't be destructable while being used in ListSupportedConversions() and IsFormatConversionSupported().
Nothing can help when the actual device/interfaces get invalidated by the system, we can only catch the errors, destruct and reconstruct later.

CrystalP committed:
[Windows][dxva] promote ID3D11VideoProcessorEnumerator1 to class member

CrystalP committed:
[Windows][dxva] dump the conversions supported by the processor for the input format in the debug log

thexai commented on xbmc/cores/VideoPlayer/VideoRenderers/HwDecRender/DXVAHD.cpp:
```diff
@@ -17,6 +17,7 @@
 #include "VideoRenderers/RenderManager.h"
 #include "VideoRenderers/windows/RendererBase.h"
 #include "rendering/dx/RenderContext.h"
+#include "utils/StringUtils.h"
```
This header is necessary?
It doesn't seem to make use of it, at least in the commit where it is "[Windows[dxva] promote ID3D11VideoProcessorEnumerator1 to class member"

Also typo in commit name `[Windows[dxva]`

thexai commented on xbmc/rendering/dx/DeviceResources.cpp:
```diff
@@ -1399,3 +1386,31 @@ DEBUG_INFO_RENDER DX::DeviceResources::GetDebugInfo() const
+std::vector<DXGI_COLOR_SPACE_TYPE> DX::DeviceResources::GetSwapChainColorSpaces() const
+{
+  if (!m_swapChain)
+    return std::vector<DXGI_COLOR_SPACE_TYPE>();
```
```suggestion
    return {};
```
Is also valid and seems more simple

thexai reviewed (approved):
Tested Windows x64 and Xbox Series S

CrystalP commented:
The use of the CriticalSection was bugging me so I checked. All functions of DXVAHD are called exclusively on the render thread (main thread) so I'm not sure what the critical section is supposed to protect. 

Did you have other concerns or does this seem ready to merge?

The Jenkins build failure of android-arm64-docker is not related.

thexai commented:
Jenkins build this please

thexai commented:
@CrystalP 
All is green, merge it whenever you want
